File: backend/app/schemas/threat.py
from enum import Enum
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field

class RiskLevel(str, Enum):
    BENIGN = "BENIGN"
    LOW_RISK = "LOW RISK"
    SUSPICIOUS = "SUSPICIOUS"
    HIGH_RISK = "HIGH RISK"
    CRITICAL = "CRITICAL"

class TargetType(str, Enum):
    URL = "URL"
    MESSAGE = "MESSAGE"
    EMAIL = "EMAIL"
    SCREENSHOT = "SCREENSHOT"

class EvidenceCategory(str, Enum):
    INFRASTRUCTURE = "INFRASTRUCTURE"
    IDENTITY = "IDENTITY"
    CONTENT = "CONTENT"
    REPUTATION = "REPUTATION"
    CRYPTOGRAPHY = "CRYPTOGRAPHY"

class EvidenceSeverity(str, Enum):
    INFO = "INFO"
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"

class EvidenceItem(BaseModel):
    category: EvidenceCategory
    severity: EvidenceSeverity
    title: str
    description: str
    technical_proof: str
    why_this_matters: str

class RedirectHop(BaseModel):
    hop_number: int
    from_url: str
    to_url: str
    status_code: int
    ip_address: Optional[str] = None
    response_time_ms: Optional[float] = None

class RecommendedAction(BaseModel):
    priority: str = Field(..., description="IMMEDIATE, PREVENTATIVE, or REPORTING")
    action: str
    rationale: str
    action_type: str = Field(..., description="DO_NOT_CLICK, CHANGE_PASSWORD, REPORT_PHISHING, VERIFY_SENDER, SAFE_TO_PROCEED")

class DNSRecords(BaseModel):
    a: List[str] = []
    aaaa: List[str] = []
    mx: List[str] = []
    ns: List[str] = []
    txt: List[str] = []

class SSLCertInfo(BaseModel):
    issuer: Optional[str] = None
    subject: Optional[str] = None
    valid_from: Optional[str] = None
    valid_until: Optional[str] = None
    is_valid: bool = False
    is_self_signed: bool = False
    days_remaining: Optional[int] = None

class TechnicalMetadata(BaseModel):
    domain: str = ""
    subdomain: str = ""
    registered_domain: str = ""
    tld: str = ""
    ip_addresses: List[str] = []
    dns: Optional[DNSRecords] = None
    ssl: Optional[SSLCertInfo] = None
    redirect_hops: List[RedirectHop] = []
    entropy: float = 0.0
    detected_brands: List[str] = []
    extracted_urls: List[str] = []
    social_engineering_flags: List[str] = []

class ThreatReport(BaseModel):
    scan_id: str
    target_type: TargetType
    raw_target: str
    defanged_target: str
    risk_score: int = Field(..., ge=0, le=100)
    risk_level: RiskLevel
    confidence_score: float = Field(..., ge=0.0, le=100.0)
    executive_summary: str
    layman_verdict: str
    evidence_items: List[EvidenceItem]
    recommended_actions: List[RecommendedAction]
    technical_metadata: TechnicalMetadata
    analyzed_at: str

class URLScanRequest(BaseModel):
    url: str
    deep_scan: bool = True

class MessageScanRequest(BaseModel):
    content: str
    sender_metadata: Optional[str] = None
    subject: Optional[str] = None
    sender: Optional[str] = None

class ScanHistoryItem(BaseModel):
    scan_id: str
    target_type: TargetType
    defanged_target: str
    risk_level: RiskLevel
    risk_score: int
    analyzed_at: str
    executive_summary: str

class StatsSummary(BaseModel):
    total_scans: int
    threats_blocked: int
    benign_verified: int
    suspicious_flagged: int
    scans_by_type: Dict[str, int]

File: frontend/src/types/threat.ts
export type RiskLevel =
  | "BENIGN"
  | "LOW RISK"
  | "SUSPICIOUS"
  | "HIGH RISK"
  | "CRITICAL";

export type TargetType = "URL" | "MESSAGE" | "EMAIL" | "SCREENSHOT";

export type EvidenceCategory =
  | "INFRASTRUCTURE"
  | "IDENTITY"
  | "CONTENT"
  | "REPUTATION"
  | "CRYPTOGRAPHY";

export type EvidenceSeverity = "INFO" | "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export interface EvidenceItem {
  category: EvidenceCategory;
  severity: EvidenceSeverity;
  title: string;
  description: string;
  technical_proof: string;
  why_this_matters: string;
}

export interface RedirectHop {
  hop_number: number;
  from_url: string;
  to_url: string;
  status_code: number;
  ip_address?: string;
  response_time_ms?: number;
}

export interface RecommendedAction {
  priority: "IMMEDIATE" | "PREVENTATIVE" | "REPORTING";
  action: string;
  rationale: string;
  action_type:
    | "DO_NOT_CLICK"
    | "CHANGE_PASSWORD"
    | "REPORT_PHISHING"
    | "VERIFY_SENDER"
    | "SAFE_TO_PROCEED";
}

export interface DNSRecords {
  a: string[];
  aaaa: string[];
  mx: string[];
  ns: string[];
  txt: string[];
}

export interface SSLCertInfo {
  issuer?: string;
  subject?: string;
  valid_from?: string;
  valid_until?: string;
  is_valid: boolean;
  is_self_signed: boolean;
  days_remaining?: number;
}

export interface TechnicalMetadata {
  domain: string;
  subdomain: string;
  registered_domain: string;
  tld: string;
  ip_addresses: string[];
  dns?: DNSRecords;
  ssl?: SSLCertInfo;
  redirect_hops: RedirectHop[];
  entropy: number;
  detected_brands: string[];
  extracted_urls: string[];
  social_engineering_flags: string[];
}

export interface ThreatReport {
  scan_id: string;
  target_type: TargetType;
  raw_target: string;
  defanged_target: string;
  risk_score: number;
  risk_level: RiskLevel;
  confidence_score: number;
  executive_summary: string;
  layman_verdict: string;
  evidence_items: EvidenceItem[];
  recommended_actions: RecommendedAction[];
  technical_metadata: TechnicalMetadata;
  analyzed_at: string;
}

export interface ScanHistoryItem {
  scan_id: string;
  target_type: TargetType;
  defanged_target: string;
  risk_level: RiskLevel;
  risk_score: number;
  analyzed_at: string;
  executive_summary: string;
}

export interface StatsSummary {
  total_scans: number;
  threats_blocked: number;
  benign_verified: number;
  suspicious_flagged: number;
  scans_by_type: Record<string, number>;
}

export type SentinelState =
  | "STANDBY"
  | "IDLE"
  | "INGESTING"
  | "SCANNING"
  | "CORRELATING"
  | "VERIFIED"
  | "ALERT";
